[Core] Abstract reference counter behind interface for more defined API

Kunchd opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
We are moving towards a goal of defining clearer interfaces between how each logical component interact with each other. This commit is meant to introduce a reference counter interface to abstract all reference counter interactions and define an agreement for what we should expect from reference counter.


<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number
N/A

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run pre-commit jobs to lint the changes in this PR. ([pre-commit setup](https://docs.ray.io/en/latest/ray-contribute/getting-involved.html#lint-and-formatting))
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Commit:
Abstract reference counter behind interface for more defined API

We are moving towards a goal of defining clearer interfaces between how each
logical component interact with each other. This commit is meant to
introduce a reference counter interface to abstract all reference counter
interactions and define an agreement for what we should expect from reference counter.

Signed-off-by: davik <davik@anyscale.com>

israbbani commented on src/ray/core_worker/reference_counter_interface.h:
```diff
@@ -0,0 +1,540 @@
+
+  //// Checks to see if objects have an owner.
+  ///
+  /// \param[in] object_id The ID of the object.
```
This looks real.

israbbani reviewed (commented):
Looking good so far!

Kunchd commented on src/ray/core_worker/reference_counter_interface.h:
```diff
@@ -0,0 +1,540 @@
+// Interface for mocking.
+class ReferenceCounterInterface {
+ public:
+  using LineageReleasedCallback =
```
It will be used by reference_counter

Commit:
Address doc related PR comments

Signed-off-by: davik <davik@anyscale.com>

Kunchd commented on src/ray/core_worker/store_provider/plasma_store_provider.h:
```diff
@@ -239,7 +239,7 @@ class CoreWorkerPlasmaStoreProvider {
   std::shared_ptr<plasma::PlasmaClientInterface> store_client_;
   /// Used to look up a plasma object's owner.
-  ReferenceCounter &reference_counter_;
+  ReferenceCounterInterface &reference_counter_;
```
Will refactor the reference counter interface out of the plasma store in a separate PR.

Kunchd commented on src/ray/core_worker/store_provider/memory_store/memory_store.h:
```diff
@@ -208,7 +208,7 @@ class CoreWorkerMemoryStore {
   /// If enabled, holds a reference to local worker ref counter. TODO(ekl) make this
   /// mandatory once Java is supported.
-  ReferenceCounter *ref_counter_;
+  ReferenceCounterInterface *ref_counter_;
```
Yep, I will refactor the reference counter interface out in a separate PR.

Commit:
Address PR comments to change dependencies to ref count interface when applicable

This commit addresses the PR comments to change dependencies from reference_counter
to reference_counter_interface when possible to speed up build times. As a result
Some tests are modified to reflect the new change in dependecies as they depend on
the actual ReferenceCounter class.

Signed-off-by: davik <davik@anyscale.com>

israbbani commented on src/ray/core_worker/tests/task_manager_test.cc:
```diff
@@ -220,7 +220,7 @@ class TaskManagerTest : public ::testing::Test {
   std::unique_ptr<MockTaskEventBuffer> task_event_buffer_mock_;
   std::shared_ptr<gcs::MockGcsClient> mock_gcs_client_;
-  std::shared_ptr<ReferenceCounter> reference_counter_;
+  std::shared_ptr<ReferenceCounterInterface> reference_counter_;
```
I think you need to include the `reference_counter_interface.h` header in all of these files. It's still working through transitive includes.

edoakes replied:
unfortunately I haven't been able to find a clean way to have bazel or a linter tool enforce this for us :(

Commit:
Address dependency style comments

Signed-off-by: davik <davik@anyscale.com>

israbbani reviewed (approved):
🚀 pinky promise for follow ups
1. delete the mock
2. try to remove the ReferenceCounter dependency of PlasmaStoreProvider and CoreWorkerMemoryStore.
